### PR TITLE
Improvements to SSH known_hosts handling

### PR DESCRIPTION
**Pull request by Shados**
### What does this PR do?
Improves SSH known_hosts handling and file management by adding and using new function variants that are aware of multiple entries for a single host, and deprecating the old single-entry-only functions.

### What issues does this PR fix or reference?
#13484
#31581
#31117

### Previous Behavior
Previously, `recv_known_host` and `get_known_host` were used internally to retrieve SSH host key information from servers and known_hosts files, respectively. Both of these functions are flawed in that they only return a single known_host entry, when there can be multiple entries for a single server (either because of having keys for multiple encryption types, or because of manually-added extra entries in a known_hosts file).

Additionally, `set_known_host` would internally use `rm_known_host` when removing existing known_hosts entries, which would remove _all_ entries for the given host/port combination, rather than only ones that match the host + port + key enc type.

### New Behavior
Now, new `recv_known_host_entries` and `get_known_host_entries` functions are used internally for the same purposes, and can return multiple known_host entries. As the old functions were public, I've marked them as deprecated rather than removing them outright, but they're no longer used internally.

`set_known_host` no longer uses `rm_known_host`, but instead parses the known_hosts file to locate and remove lines with a match host + port + key enc type.

### Tests written?

No.

I've minimally modified the existing tests to pass against the new functions instead of the deprecated ones, but it would be nice to extend the test cases to cover the multiple-known-host-entries functionality. 

However, the current tests are done by checking against github.com, and github only uses an RSA host key, so it would not be useable for these tests. I'm also not convinced it's all that sensible to have a remote dependency during testing unless it's absolutely necessary. Similarly, it'd be nice to have regression tests for the fixed issues, but this runs into the same issue.

-------

I've read the contributing guide, but I may have missed something, so let me know if so. Two things related to this:
- I'm not sure if this PR is more appropriate against the develop branch or the current release branch... it is a bug fix, but it's a bug in the underlying design of the functions rather than in their implementation per se, and so does require the new functions to avoid breaking code dependent on the old semantics.
- I don't know what should be put down for `versionadded` on the new function variants.

### CODE REVIEW

File: salt/modules/ssh.py [hunk 1]
```diff
@@ -834,6 +848,32 @@ def get_known_host(user,
...
+    return known_hosts[0] if known_hosts else None
+
+@salt.utils.decorators.path.which('ssh-keygen')
+def get_known_host_entries(user,
+                           hostname,
+                           config=None,
+                           port=None,
+                           fingerprint_hash_type=None):
+    '''
+    .. versionadded:: 201?.?.?
```
Comment (cachedout) [inline salt/modules/ssh.py]: This should be `Oxygen`

File: salt/modules/ssh.py [hunk 2]
```diff
@@ -897,6 +935,59 @@ def recv_known_host(hostname,
...
+
+@salt.utils.decorators.path.which('ssh-keyscan')
+def recv_known_host_entries(hostname,
+                            enc=None,
+                            port=None,
+                            hash_known_hosts=True,
+                            timeout=5,
+                            fingerprint_hash_type=None):
+    '''
+    .. versionadded:: 201?.?.?
```
Comment (cachedout) [inline salt/modules/ssh.py]: `Oxygen`

**Review by cachedout — CHANGES_REQUESTED** [review]

**Comment by Shados** [comment]
Updated with the versionadded and linter fixes (I think).

**Review by cachedout — APPROVED** [review]

File: salt/modules/ssh.py [hunk 3]
```diff
@@ -837,6 +853,33 @@ def get_known_host(user,
 
         salt '*' ssh.get_known_host <user> <hostname>
     '''
+    salt.utils.warn_until(
+            'Neon',
+            'get_known_host has been deprecated in favour of '
+            'get_known_host_entries.'
```
Comment (rallytime) [inline salt/modules/ssh.py]: Can you add a mention to this warning about when the function will be removed? Something like `\'get_known_host\' has been deprecated in favour of \'get_known_host_entries\'. \'get_known_host\' will be removed in Salt Neon.'`.

That will help users know when they need to upgrade by.

File: salt/modules/ssh.py [hunk 4]
```diff
@@ -900,6 +942,60 @@ def recv_known_host(hostname,
 
         salt '*' ssh.recv_known_host <hostname> enc=<enc> port=<port>
     '''
+    salt.utils.warn_until(
+            'Neon',
+            'recv_known_host has been deprecated in favour of '
+            'recv_known_host_entries.'
```
Comment (rallytime) [inline salt/modules/ssh.py]: Same comment for this deprecation warning as well. :)

File: salt/modules/ssh.py [hunk 5]
```diff
@@ -837,6 +853,33 @@ def get_known_host(user,
 
         salt '*' ssh.get_known_host <user> <hostname>
     '''
```
Comment (rallytime) [inline salt/modules/ssh.py]: Can you add a `.. deprecated:: Oxygen` tag to the doc-string for this function?

File: salt/modules/ssh.py [hunk 6]
```diff
@@ -900,6 +942,60 @@ def recv_known_host(hostname,
 
         salt '*' ssh.recv_known_host <hostname> enc=<enc> port=<port>
     '''
```
Comment (rallytime) [inline salt/modules/ssh.py]: This function also needs a `.. deprecated:: Oxygen` tag here.

**Review by rallytime — COMMENTED** [review]
This looks good overall, I just had a couple of small comments. :)

**Comment by rallytime** [comment]
Hi @Shados - did you get a chance to review my comments above? Once those are fixed up we can get this merged in.

**Comment by Shados** [comment]
@rallytime yep, just fairly busy at the moment. Will probably get around to it on Tuesday or so.

**Comment by rallytime** [comment]
@Shados That sounds great. Thank you!

**Review by rallytime — APPROVED** [review]

**Comment by rallytime** [comment]
Thanks again for submitting this @Shados and welcome to the Salt community! I made a quick commit to this PR to resolve minor comments I left before as we'd like to get this over the line. :) Once tests are done running, I'll get this merged in.

**Comment by Shados** [comment]
@rallytime Ah, thanks for that. I ended up having to fight a bunch of fires at work before going on leave for a while, so this got pushed way back :).

**Pull request merged**